x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -490,7 +492,87 @@ static XContentBuilder buildUpdatedDocument(
 
         addCreator(builder, authentication);
 
-        return builder.endObject();
+        return new ApiKeyDocBuilderWithNoopFlag(
+            builder.endObject(),
+            isNoop(currentApiKeyDoc, targetDocVersion, authentication, request, userRoles)
+        );
```

Comment:
This works. But I think it can be simplified a bit if:
1. Perform `isNoop` check in the beginning of the method. So we can avoid creating and preparing the builder all together if no change.
2. Remove the `ApiKeyDocBuilderWithNoopFlag` record type and just return `null` if there is no change or the actual builder if there are any changes.

Reply:
Performing the noop check at the beginning of the method is how I had this structured initially, so I'm happy to switch it back to that. My thinking around always building was that it makes the method more generic and gives it simpler semantics: it creates the object and tells the caller if it's a noop -- then it's up to the caller to decide what to do with it.

This is just to give context. I'm good with your suggestion 👍